src/main/java/minijava/ir/utils/FirmUtils.java

```diff
@@ -0,0 +1,46 @@
+package minijava.ir.utils;
+
+import firm.nodes.Address;
+import firm.nodes.Block;
+import firm.nodes.Node;
+import firm.nodes.Phi;
+import java.util.*;
+
+public class FirmUtils {
+
+  private static Map<Phi, Boolean> isPhiProneToLostCopies = new HashMap<>();
+
+  public static String getMethodLdName(firm.nodes.Call node) {
+    return ((Address) node.getPred(1)).getEntity().getLdName();
+  }
+
+  public static boolean isPhiProneToLostCopies(Phi phi) {
+    if (!isPhiProneToLostCopies.containsKey(phi)) {
+      isPhiProneToLostCopies.put(phi, checkPhi(phi));
+    }
+    return isPhiProneToLostCopies.get(phi);
+  }
```

Comment:
Well, for one it assumes that the predecessors of the cached `Phi`s aren't changed between invocations of `isPhiProneToLostCopies`. Meaning that if at one point, we decide to implement an optimization which might be able to optimize some a `Phi` after it is sure that the `Phi` `isPhiProneToLostCopies`, the results might change. In other words: You memoize a function which is not pure, producing different output depending on where/when you call it.

I'd move the caching component completely into the `AssemblyGenerator`, which is the only place where you can maybe live with that invariant. `checkPhi` may stay in this public helper, but the caching wrapper is pretty much unsafe to use outside of the `AssemblyGenerator`.